src/journal/test-journal-append.c

```diff
@@ -0,0 +1,273 @@
+#include "tmpfile-util.h"
+#include "tests.h"
+
+static int journal_append_message(ManagedJournalFile *mj, const char *message) {
+        struct iovec iovec;
+        struct dual_timestamp ts;
+
+        assert(mj);
+        assert(message);
+
+        dual_timestamp_get(&ts);
+        iovec = IOVEC_MAKE_STRING(message);
+        return journal_file_append_entry(
+                                mj->file,
+                                &ts,
+                                /* boot_id= */ NULL,
+                                &iovec,
+                                /* n_iovec= */ 1,
+                                /* seqnum= */ NULL,
+                                /* seqnum_id= */ NULL,
+                                /* ret_object= */ NULL,
+                                /* ret_offset= */ NULL);
+}
+
+static int journal_corrupt_and_append(uint64_t start_offset, uint64_t step) {
+        _cleanup_(mmap_cache_unrefp) MMapCache *mmap_cache = NULL;
+        _cleanup_(rm_rf_physical_and_freep) char *tempdir = NULL;
+        _cleanup_(managed_journal_file_closep) ManagedJournalFile *mj = NULL;
+        uint64_t start, end;
+        int r;
+
+        mmap_cache = mmap_cache_new();
+        assert_se(mmap_cache != NULL);
+
+        /* managed_journal_file_open() requires a valid machine id */
+        if (sd_id128_get_machine(NULL) < 0)
+                return log_tests_skipped("No valid machine ID found");
+
+        assert_se(mkdtemp_malloc("/tmp/journal-append-XXXXXX", &tempdir) >= 0);
+        assert_se(chdir(tempdir) >= 0);
+        (void) chattr_path(tempdir, FS_NOCOW_FL, FS_NOCOW_FL, NULL);
+
+        log_debug("Opening journal %s/system.journal", tempdir);
+
+        r = managed_journal_file_open(
+                                /* fd= */ -1,
+                                "system.journal",
+                                O_RDWR|O_CREAT,
+                                JOURNAL_COMPRESS,
+                                0644,
+                                /* compress_threshold_bytes= */ UINT64_MAX,
+                                /* metrics= */ NULL,
+                                mmap_cache,
+                                /* deferred_closes= */ NULL,
+                                /* template= */ NULL,
+                                &mj);
+        if (r < 0)
+                return log_error_errno(r, "Failed to open the journal: %m");
+
+        assert_se(mj);
+        assert_se(mj->file);
+
+        /* Add a couple of initial messages */
+        for (int i = 0; i < 10; i++) {
+                _cleanup_free_ char *message = NULL;
+
+                assert_se(asprintf(&message, "MESSAGE=Initial message %d", i));
+                r = journal_append_message(mj, message);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to write to the journal: %m");
+        }
+
+        start = start_offset == UINT64_MAX ? random_u32() % mj->file->last_stat.st_size : start_offset;
```

Comment:
Not a matter, but why not `random_u64()` ?

Reply:
No idea to be honest :)